File: frontend/src/types/recommendation.ts
export interface CompositeRecommendation {
  rank: number;
  crop_id: string;
  crop_name: string;
  category: string;
  composite_suitability_score: number;
  ml_confidence_pct: number;
  ecocrop_score: number;
  fao_suitability_class: string;
  estimated_yield_ton_ha: number;
  estimated_gross_revenue_usd_ha: number;
  estimated_cultivation_cost_usd_ha: number;
  estimated_net_profit_usd_ha: number;
  economic_roi_pct: number;
  risk_level: string;
  primary_limiting_factor: string;
  agronomic_advisory: string;
}

export interface RecommendationRequest {
  farm_id?: string;
  field_parcel_id?: string;
  soil_sample_id?: string;
  n_kg_ha: number;
  p_kg_ha: number;
  k_kg_ha: number;
  ph: number;
  organic_carbon_pct: number;
  ec_ds_m: number;
  soil_texture: string;
  temperature_c: number;
  temp_max_c?: number;
  temp_min_c?: number;
  humidity_pct: number;
  rainfall_mm: number;
  elevation_m: number;
  category_filter?: string;
  top_k: number;
  target_budget_usd_ha?: number;
}

export interface RecommendationDetailResponse {
  crop_rankings: CompositeRecommendation[];
  top_crop_fertilizer_schedule: any;
  top_crop_yield_risk_analysis: any;
  top_crop_pest_alerts: any;
  ensemble_metadata: Record<string, number>;
  query_timestamp: string;
}

export interface WhatIfScenarioRequest {
  base_request: RecommendationRequest;
  simulated_rainfall_delta_pct: number;
  simulated_temperature_delta_c: number;
  supplemental_irrigation_mm: number;
  additional_fertilizer_budget_pct: number;
}

export interface WhatIfScenarioResponse {
  baseline_top_crop: string;
  baseline_suitability: number;
  simulated_top_crop: string;
  simulated_suitability: number;
  rank_changes: Array<{
    crop_id: string;
    crop_name: string;
    new_rank: number;
    new_suitability_score: number;
  }>;
  water_stress_shift_pct: number;
  summary_insight: string;
}

File: backend/app/schemas/recommendation.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from app.ml.pipeline.ensemble import CompositeRecommendation
from app.ml.agronomy.ssnm_fertilizer import FertilizerRecommendationSchedule
from app.ml.agronomy.yield_estimator import YieldEstimationReport
from app.ml.agronomy.pest_risk_index import CropPestRiskReport


class RecommendationRequest(BaseModel):
    farm_id: Optional[str] = None
    field_parcel_id: Optional[str] = None
    soil_sample_id: Optional[str] = None

    # Soil Parameters (either supplied or inferred from soil_sample_id)
    n_kg_ha: float = Field(280.0, ge=0)
    p_kg_ha: float = Field(30.0, ge=0)
    k_kg_ha: float = Field(180.0, ge=0)
    ph: float = Field(6.8, ge=3.0, le=12.0)
    organic_carbon_pct: float = Field(0.55, ge=0.0)
    ec_ds_m: float = Field(0.7, ge=0.0)
    soil_texture: str = "loam"

    # Meteorological / Climate Parameters (either supplied or fetched via coordinates)
    temperature_c: float = Field(26.0, ge=-10.0, le=55.0)
    temp_max_c: Optional[float] = None
    temp_min_c: Optional[float] = None
    humidity_pct: float = Field(65.0, ge=5.0, le=100.0)
    rainfall_mm: float = Field(750.0, ge=0.0)
    elevation_m: float = 150.0

    # Operational Preferences
    category_filter: Optional[str] = None  # Cereals, Pulses, Oilseeds, Commercial, Vegetables, Fruits, etc.
    top_k: int = Field(10, ge=1, le=50)
    target_budget_usd_ha: Optional[float] = None


class RecommendationDetailResponse(BaseModel):
    crop_rankings: List[CompositeRecommendation]
    top_crop_fertilizer_schedule: FertilizerRecommendationSchedule
    top_crop_yield_risk_analysis: YieldEstimationReport
    top_crop_pest_alerts: CropPestRiskReport
    ensemble_metadata: Dict[str, Any]
    query_timestamp: datetime


class WhatIfScenarioRequest(BaseModel):
    base_request: RecommendationRequest
    simulated_rainfall_delta_pct: float = Field(0.0, description="% change in rainfall (-50 to +100%)")
    simulated_temperature_delta_c: float = Field(0.0, description="Temperature shift (-5 to +5°C)")
    supplemental_irrigation_mm: float = Field(0.0, ge=0.0)
    additional_fertilizer_budget_pct: float = Field(0.0)


class WhatIfScenarioResponse(BaseModel):
    baseline_top_crop: str
    baseline_suitability: float
    simulated_top_crop: str
    simulated_suitability: float
    rank_changes: List[Dict[str, Any]]
    water_stress_shift_pct: float
    summary_insight: str
